xwiki-platform-core/xwiki-platform-resource/xwiki-platform-resource-temporary/src/main/java/org/xwiki/resource/temporary/internal/DefaultTemporaryResourceStore.java

```diff
@@ -75,31 +76,35 @@ public File createTemporaryFile(TemporaryResourceReference reference, InputStrea
     public File getTemporaryFile(TemporaryResourceReference reference) throws IOException
     {
         List<String> segments = new ArrayList<>();
-        segments.add("tmp");
-        segments.add(reference.getModuleId());
-        int safePathLength = 2;
+
         if (reference.getOwningEntityReference() != null) {
             for (EntityReference component : reference.getOwningEntityReference().getReversedReferenceChain()) {
                 segments.add(component.getName());
-                safePathLength++;
             }
         }
         if (!reference.getParameters().isEmpty()) {
             segments.add(String.valueOf(reference.getParameters().hashCode()));
-            safePathLength++;
         }
-        segments.addAll(reference.getResourcePath());
+
         String path = StringUtils.join(encode(segments), '/');
-        String safePath = StringUtils.join(encode(segments.subList(0, safePathLength)), '/');
+        String md5 = DigestUtils.sha1Hex(path);
+
+        List<String> finalPathSegments = new ArrayList<>();
+
+        finalPathSegments.add("tmp");
+        finalPathSegments.add(reference.getModuleId());
+
+        // Avoid having too many files in one folder because some filesystems don't perform well with large numbers of
+        // files in one folder
+        finalPathSegments.add(String.valueOf(md5.charAt(0)));
+        finalPathSegments.add(String.valueOf(md5.charAt(1)));
+        finalPathSegments.add(String.valueOf(md5.substring(2)));
+
+        finalPathSegments.addAll(reference.getResourcePath());
+
         File rootFolder = this.environment.getTemporaryDirectory();
-        File safeFolder = new File(rootFolder, safePath);
-        File temporaryFile = new File(rootFolder, path);
 
-        // Make sure the resource path is not relative (e.g. "../../../") and tries to get outside of the safe folder.
-        if (!temporaryFile.getAbsolutePath().startsWith(safeFolder.getAbsolutePath())) {
```

Comment:
Why have you removed this protection ? It's still possible to have this kind of things in the resource path, so there is no reason to remove that.

Reply:
Added safe check